packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -138,8 +150,42 @@ export class BitbucketUrlReader implements UrlReader {
     throw new Error(message);
   }
 
-  readTree(): Promise<ReadTreeResponse> {
-    throw new Error('BitbucketUrlReader does not implement readTree');
+  async readTree(
+    url: string,
+    options?: ReadTreeOptions,
+  ): Promise<ReadTreeResponse> {
+    const gitUrl: parseGitUri.GitUrl = parseGitUri(url);
+    const { name: repoName, owner, ref, protocol, resource, filepath } = gitUrl;
+
+    const isHosted = resource === 'bitbucket.org';
+
+    if (isHosted && !ref) {
+      // TODO(freben): We should add support for defaulting to the default branch
+      throw new InputError(
+        'Bitbucket URL must contain branch to be able to fetch tree',
+      );
+    }
+
+    const archiveUrl = isHosted
+      ? `${protocol}://${resource}/${owner}/${repoName}/get/${ref}.zip`
+      : `${protocol}://${resource}/projects/${owner}/repos/${repoName}/archive?format=zip`;
+
+    const response = await fetch(archiveUrl, getApiRequestOptions(this.config));
+    if (!response.ok) {
+      const message = `Failed to read tree from ${url}, ${response.status} ${response.statusText}`;
+      if (response.status === 404) {
+        throw new NotFoundError(message);
+      }
+      throw new Error(message);
+    }
+
+    const lastCommitShortHash = await this.getLastCommitShortHash(gitUrl);
+
+    return this.treeResponseFactory.fromZipArchive({
+      stream: (response.body as unknown) as Readable,
+      path: `${owner}-${repoName}-${lastCommitShortHash}/${filepath}`,
```

Comment:
@Rugvip It would be convenient if the response factory could unconditionally skip over levels e.g. using blobs so we could make this `*/${filepath}`. I'll make a ticket.

Reply:
Yep, let's merge without that addition but agree that it might be nice. Should just be a matter of doing the `subPath` slicing slightly differently, I don't think we want to support anything but prefixed `*/` though, and possibly even making the API be a `depth` option rather than communicating it as part of the path.

Reply:
Ping @Rugvip @freben Did you have a ticket already? We should at least extract the directory name from the archive instead of guessing it for all different types of providers. Currently, this is wrong from Bitbucket hosted server (https://github.com/backstage/backstage/issues/4121) and the bug can be avoidable if we don't hardcode the name of the "unzipped" directory.

I'll create an issue, if you haven't already. 🙏

Reply:
No worries, I created a PR instead! https://github.com/backstage/backstage/pull/4169